# Sentiment Analysis Using Bag of Words and Decision Tree

## Project Overview

Natural Language Processing (NLP) enables machines to understand, analyze, and extract meaningful information from human language. One of the most fundamental NLP tasks is **Sentiment Analysis**, which aims to determine whether a piece of text expresses a positive or negative opinion.

In this project, I build a complete sentiment analysis pipeline using classical NLP techniques and Machine Learning algorithms. The workflow covers the entire process from raw text preprocessing to feature extraction, model training, evaluation, and experiment documentation.

The primary objective of this project is not only to achieve high predictive performance but also to systematically investigate how different preprocessing techniques, feature engineering strategies, and machine learning models affect sentiment classification performance.

---

## Dataset

Dataset Link: **[IMBD Movie Reviews](https://www.kaggle.com/datasets/mwallerphunware/imbd-movie-reviews-for-binary-sentiment-analysis)**

The dataset consists of movie reviews labeled with their corresponding sentiment:

* Positive
* Negative

The reviews contain real-world natural language, making them suitable for evaluating text preprocessing techniques and machine learning models.

---

## Project Pipeline

The following stages are implemented throughout this project:

### 1. Text Preprocessing

* Contraction expansion
* Lowercasing
* Text cleaning using regular expressions
* Stopword removal with preserved negations
* Lemmatization
* Corpus construction

### 2. Feature Extraction

* Bag of Words (CountVectorizer)
* N-gram generation
* Vocabulary analysis
* Feature selection using `max_features`
* Rare-word filtering using `min_df`

### 3. Model Training

Different machine learning algorithms are evaluated and compared, including:

* Gaussian Naive Bayes
* Multinomial Naive Bayes
* Bernoulli Naive Bayes
* Logistic Regression
* Support Vector Machines (SVM)
* Decision Trees
* Random Forests

### 4. Model Evaluation

Performance is assessed using multiple metrics:

* Accuracy
* Precision
* Recall
* F1-Score
* ROC-AUC Score
* Confusion Matrix
* Cross-Validation Mean Accuracy
* Cross-Validation Standard Deviation

---

## Experimental Approach

Rather than training a single model, this notebook follows an experimentation-driven methodology.

For each model, multiple configurations are tested, including different values for:

* `max_features`
* `min_df`
* N-gram ranges
* Preprocessing strategies

The goal is to identify the most effective configuration while understanding the trade-offs between model complexity, computational cost, and predictive performance.

---

## Key Learning Objectives

Through this project, I aim to:

* Develop a deeper understanding of NLP preprocessing techniques.
* Compare the behavior of different machine learning algorithms on text data.
* Analyze how feature engineering impacts classification performance.
* Build reproducible NLP pipelines suitable for real-world applications.
* Establish strong baselines before moving toward advanced embedding and transformer-based approaches.

---

**Author:** Hazem Mohamed

**Role:** AI Engineer | Machine Learning Engineer | NLP Engineer

**Repository:** [NLP Experimentation Lab](https://github.com/Hazem1695/NLP-Experimentation-Lab)


# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

## **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv('MovieReviewTrainingDatabase.csv')
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  25000 non-null  object
 1   review     25000 non-null  object
dtypes: object(2)
memory usage: 390.8+ KB
None

SHAPE OF DATA
Rows: 25000, Columns: 2

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 96

DATA TYPES
sentiment    object
review       object
dtype: object

NUMERICAL SUMMARY
       sentiment                                             review
count      25000                                              25000
unique         2                                              24904
top     Positive  You do realize that you've been watching the E...
freq       12500                                                  3

CATEGORICAL SUMMARY
       sentiment                                             review
count      25000                 

## Duplicate Data Detection

In [4]:
duplicates = dataset[dataset.duplicated(subset=['review'], keep=False)]
duplicates.sort_values('review')

,sentiment,review
21186,Negative,"Back in his youth, the old man had wanted to..."
21877,Negative,"Back in his youth, the old man had wanted to..."
14734,Negative,'Dead Letter Office' is a low-budget film abou...
5519,Negative,'Dead Letter Office' is a low-budget film abou...
7011,Positive,".......Playing Kaddiddlehopper, Col San Fernan..."
...,...,...
2685,Negative,"in this movie, joe pesci slams dunks a basketb..."
22244,Positive,it's amazing that so many people that i know h...
14767,Positive,it's amazing that so many people that i know h...
12462,Negative,this movie begins with an ordinary funeral... ...


## Quantifying Duplicate Review Frequencies

In [5]:
review_counts = dataset['review'].value_counts()
print("Reviews appearing more than once:")
print((review_counts > 1).sum())
print("\nMaximum repetitions:")
print(review_counts.max())

Reviews appearing more than once:
92

Maximum repetitions:
3


## Removing Duplicate Reviews & Resetting Index
> **Note:** This cell drops the repeated rows we identified in the previous steps and cleanly resets the row indices for model training

In [6]:
print("Before:", len(dataset))
dataset = dataset.drop_duplicates()
print("After:", len(dataset))
dataset = dataset.reset_index(drop=True)

Before: 25000
After: 24904


## Library Installation
> **Note:** The `contractions` library is required to automatically expand shortcuts like *don't* to *do not* and *I'm* to *I am* during preprocessing.

In [7]:
!pip install contractions

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 11.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 5.8 MB/s eta 0:00:00


## **Cleaning the texts**

In [8]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')
import re
import contractions
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

base_stopwords = set(stopwords.words('english'))
negation_words = {'not', 'no', 'never'}
all_stopwords = base_stopwords - negation_words

wnl = WordNetLemmatizer()

corpus = []

for i in range(0, len(dataset)):
    review = dataset['review'][i]
    # Fix contractions
    review = contractions.fix(review)
    # Lowercase
    review = review.lower()
    
    review = re.sub(r'[^a-zA-Z\s]', ' ', review)
    # Split
    words = review.split()
    # Chained Lemmatization (Handles both Verbs 'v' and Nouns 'n')
    review = [wnl.lemmatize(wnl.lemmatize(word, pos='v'), pos='n') for word in words if word not in all_stopwords]
    review = ' '.join(review) 
    corpus.append(review)

[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /usr/share/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /usr/share/nltk_data...


## Preprocessing Verification
> **Note:** Pulling the first two rows directly as a memory array to confirm that our lowercasing, stopword stripping, and lemmatization pipeline worked correctly before feeding it into the vectorizer.

In [9]:
# Pull the data directly as a fast memory array
raw_samples = dataset['review'].head(2).values

for i in range(2):
    print(f"=== REVIEW #{i+1} ===")
    print(f"RAW:     {raw_samples[i]}\n") 
    print(f"CLEANED: {corpus[i]}")
    print("-" * 50)

=== REVIEW #1 ===
RAW:     With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.  Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.  The actual feature film bit when it final

# **Encoding Categorical data Using Label Encoding**

In [10]:
from sklearn.preprocessing import LabelEncoder
y = dataset.iloc[:, 0].values
le = LabelEncoder()
y = le.fit_transform(y)

In [11]:
print(y)

[1 1 0 ... 0 0 1]


# Class Balance Check
> **Note:** Using NumPy to verify if our dataset is perfectly balanced between positive and negative reviews before splitting it into training and testing sets.

In [12]:
# This returns the unique classes and how many times they appear
classes, counts = np.unique(y, return_counts=True)
for c, count in zip(classes, counts):
    print(f"Class {c} contains {count}")

Class 0 contains 12432
Class 1 contains 12472


# **Splitting the dataset into the Training set and Test set**

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(corpus, y, test_size = 0.20, random_state = 0)

# **Creating the Bag of Word model**

In [14]:
from sklearn.feature_extraction.text import CountVectorizer
# Add ngram_range=(1, 2) so it automatically catches phrases like "not good" or "no clue"
cv = CountVectorizer(max_features=5000, ngram_range=(1,2)) # first try without (max_features)
X_train = cv.fit_transform(X_train)
X_test = cv.transform(X_test)

In [15]:
print(X_train.shape)        # (n_samples, n_features)
print(X_train.shape[0])     # number of training samples
print(X_train.shape[1])     # number of features (vocabulary size)

(19923, 5000)
19923
5000


In [16]:
print(X_test.shape)        # (n_samples, n_features)
print(X_test.shape[0])     # number of training samples
print(X_test.shape[1])     # number of features (vocabulary size)

(4981, 5000)
4981
5000


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(random_state=0)

param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'splitter': ['best', 'random'],
    'max_depth': [None, 5, 10, 20, 30, 50],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 4, 8],
    'max_features': [None, 'sqrt', 'log2']
}

In [ ]:
random_search = RandomizedSearchCV(
    estimator=dt,
    param_distributions=param_grid,
    n_iter=30,          # Number of random combinations
    scoring='accuracy',
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2
)

In [20]:
random_search.fit(X_train, y_train)

print(random_search.best_params_)
print(random_search.best_score_)

Fitting 5 folds for each of 30 candidates, totalling 150 fits
{'splitter': 'random', 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_features': None, 'max_depth': 20, 'criterion': 'gini'}
0.7367867328788177


# **Training theDecision Tree model on the Training set**

In [17]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(splitter='random', max_depth=20, min_samples_split=2, min_samples_leaf=2, criterion='gini', random_state=0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=20, min_samples_leaf=2, random_state=0,
                       splitter='random')

# **Predicting the Test set results**

In [18]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [1 0]
 ...
 [0 0]
 [0 0]
 [1 0]]


# **Evaluating the Model Performance**

In [19]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=3)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())

Confusion Matrix:
[[1735  785]
 [ 493 1968]]

Accuracy Score:
0.7434250150572175

Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.69      0.73      2520
           1       0.71      0.80      0.75      2461

    accuracy                           0.74      4981
   macro avg       0.75      0.74      0.74      4981
weighted avg       0.75      0.74      0.74      4981


ROC-AUC Score:
0.7440834961913791

Mean Accuracy:
0.7296591878733123

Standard Deviation:
0.0059032035046887215


# Decision Tree Performance Analysis for Text Classification

# 1. Objective

The objective of this experiment was to evaluate the effectiveness of the **Decision Tree** algorithm for binary text classification using **Bag-of-Words (BoW)** features generated by **CountVectorizer**.

Unlike linear models and probabilistic classifiers, Decision Trees partition the feature space through a sequence of hierarchical decisions. Although Decision Trees are highly interpretable, they are often challenged by high-dimensional and sparse text representations. This study investigates how feature extraction, tree complexity, and hyperparameter optimization influence their performance.

The experiment aims to answer the following research questions:

* How well does Decision Tree perform on sparse text data?
* Does increasing the vocabulary size always improve performance?
* Can hyperparameter optimization significantly improve the classifier?
* Why was **RandomizedSearchCV** selected instead of **GridSearchCV**?
* What are the limitations of Decision Trees for text classification?

---

# 2. Experimental Setup

## Dataset

* Final test set: **4,981 documents** (2,520 negative / 2,461 positive).

## Feature Extraction

Documents were transformed into numerical vectors using **CountVectorizer**.

| Configuration | Parameters                            |
| -------------- | -------------------------------------- |
| Baseline       | `CountVectorizer()` (default settings) |
| C1             | max_features=5000, ngram_range=(1,2)   |
| C2             | max_features=10000, ngram_range=(1,2)  |

---

# 3. Hyperparameter Optimization Strategy

**GridSearchCV** was considered first for hyperparameter optimization.

The complete search space contained:

* **1,728 parameter combinations**
* **5-fold Cross Validation**

Resulting in:

```text
1,728 × 5 = 8,640 model trainings
```

Training thousands of Decision Trees on high-dimensional text data would require a considerable amount of computation time.

Instead, **RandomizedSearchCV** was selected.

Configuration:

```text
30 random parameter combinations
5-fold Cross Validation
```

Resulting in only:

```text
30 × 5 = 150 model trainings
```

This reduced the computational cost by approximately **98%** (150 vs. 8,640 fits) while exploring only **~1.7%** of the full grid — RandomizedSearchCV was chosen because it is substantially faster than GridSearchCV while still surfacing strong configurations.

---

# 4. Hyperparameters Explored

The optimization process searched over parameters including:

* Tree depth (`max_depth`)
* Splitting strategy (`splitter`)
* Splitting criterion (`gini`, `entropy`)
* Minimum samples required for splitting (`min_samples_split`)
* Minimum samples required per leaf (`min_samples_leaf`)
* Feature selection strategy (`max_features`)

**Notably, RandomizedSearchCV converged on the exact same best configuration for all three vocabulary sizes (5K, 10K, 15K):**

```python
DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2,
    max_features=None
)
```

This consistency suggests the search space has a stable optimum for this dataset that is largely independent of vocabulary size — the tree's own capacity constraints (depth, leaf/split minimums), rather than the number of available features, appear to be the dominant factor in performance.

---

# 5. Experimental Results

| Configuration              | Accuracy   | ROC-AUC    | CV Mean | CV Std |
| --------------------------- | ---------- | ---------- | ------- | ------ |
| Baseline (default, no tune) | 71.37%     | 0.7137     | 70.84%  | 0.0052 |
| RandomizedSearch (5K)       | **74.34%** | **0.7441** | 72.97%  | 0.0059 |
| RandomizedSearch (10K)      | 74.08%     | 0.7412     | 72.84%  | 0.0028 |

---

# 6. Performance Analysis

## Effect of Hyperparameter Optimization

Hyperparameter tuning produced a clear improvement over the untuned baseline.

Accuracy increased from:

**71.37%** (baseline)

to

**74.34%** (best tuned model, 5K features)

representing an improvement of nearly **3 percentage points**, along with a tighter, more consistent cross-validation spread.

## Why RandomizedSearchCV?

| Method             | Candidates | Total Fits |
| ------------------- | ---------: | ---------: |
| GridSearchCV        |      1,728 |      8,640 |
| RandomizedSearchCV  |         30 |        150 |

RandomizedSearchCV evaluated only **1.7%** of the total parameter combinations yet discovered a configuration that meaningfully outperformed the untuned baseline, at roughly **2% of the computational cost** of an exhaustive grid search.

## Effect of Vocabulary Size

| Features | Accuracy   |
| -------- | ---------- |
| **5K**   | **74.34%** |
| 10K      | 74.08%     |

Unlike in prior BoW experiments (e.g. Naive Bayes, LinearSVC), where accuracy tended to improve or plateau as vocabulary size increased, **the 5K-feature configuration slightly outperformed the 10K-feature configuration here (74.34% vs. 74.08%)**. This is a notable reversal of the "more features → better" pattern.

A likely explanation is that a single Decision Tree's capacity is bounded by its depth and leaf constraints (`max_depth=20`, `min_samples_leaf=2`) rather than by the size of the input vocabulary — beyond a certain point, additional features mostly add sparse, rarely-used splitting candidates that dilute the search rather than improve it. This reinforces the broader conclusion (Section 12) that Decision Trees do not scale their accuracy with vocabulary size the way linear models do.

## Effect of Splitting Strategy

The optimized models consistently selected:

```python
splitter='random'
```

rather than `splitter='best'`. Random splitting introduces additional randomness during tree construction, which reduces overfitting and tends to generalize better on high-dimensional sparse data, where greedily choosing the locally optimal split can produce overly specific decision boundaries.

---

# 7. Precision and Recall Analysis

### Baseline (untuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.72      | 0.71   | 0.72     |
| 1     | 0.71      | 0.71   | 0.71     |

### Best Model — 5K Features (Tuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.78      | 0.69   | 0.73     |
| 1     | 0.71      | 0.80   | 0.75     |

### 10K Features (Tuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.76      | 0.71   | 0.74     |
| 1     | 0.72      | 0.77   | 0.75     |

Both tuned configurations show the same asymmetry: higher precision on Class 0 paired with higher recall on Class 1. This means the tree is more conservative about labeling something Class 0 (fewer false Class-0 calls, but it misses some true Class 0s) while being more liberal about labeling Class 1 (catches most true Class 1s, at the cost of some false positives). The 5K model shows this imbalance more sharply than the 10K model, whose class-wise metrics are more balanced despite its slightly lower overall accuracy.

---

# 8. ROC-AUC Analysis

| Configuration           | ROC-AUC    |
| ------------------------ | ---------- |
| Baseline (untuned)       | 0.7137     |
| **RandomizedSearch (5K)**| **0.7441** |
| RandomizedSearch (10K)   | 0.7412     |

ROC-AUC tracks accuracy closely across configurations, confirming that the 5K model's advantage isn't an artifact of the accuracy threshold alone — it separates the two classes slightly better overall.

---

# 9. Cross-Validation Analysis

| Configuration           | CV Mean Accuracy | CV Std Dev |
| ------------------------ | ----------------- | ---------- |
| Baseline (untuned)       | 70.84%             | 0.0052     |
| **RandomizedSearch (5K)**| **72.97%**         | 0.0059     |
| RandomizedSearch (10K)   | 72.84%             | **0.0028** |

Interestingly, while 5K has the higher CV mean accuracy, **10K shows a notably tighter standard deviation (0.0028 vs. 0.0059)** — meaning the 10K model was more *consistent* across folds even though it was slightly less accurate on average. Depending on priorities, this makes 10K arguably the more "stable" choice, while 5K is the higher-performing one on this particular test split. This is a useful nuance to note rather than declaring a single model an unambiguous winner.

---

# 10. Best Model Configuration

The highest-performing evaluated Decision Tree was:

```python
CountVectorizer(
    max_features=5000,
    ngram_range=(1, 2)
)

DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2,
    random_state=0
)
```

Performance:

* Accuracy = **74.34%**
* Precision (Class 0 / 1) = **0.78 / 0.71**
* Recall (Class 0 / 1) = **0.69 / 0.80**
* ROC-AUC = **0.7441**
* Cross-Validation Accuracy = **72.97%**
* Cross-Validation Standard Deviation = **0.0059**

---

# 11. Comparison with Previously Evaluated Models

| Model                   | Best Accuracy | ROC-AUC |
| ------------------------ | -------------- | ------- |
| Logistic Regression      | *88.78%*      | *0.8880* |
| LinearSVC                | *87.51%*      | *0.8753* |
| Naive Bayes (Multinomial) | *87.19%* | *0.8719* |
| K-Nearest Neighbors      | *76.79%*      | *0.7671* |
| **Decision Tree**        | **74.34%**     | **0.7441** |

---

# 12. Discussion

The experiments confirm the general limitation of a single Decision Tree for sparse text classification: axis-aligned splits struggle to build compact, generalizable boundaries over thousands of mostly-zero BoW dimensions.

RandomizedSearchCV again proved to be a practical substitute for GridSearchCV, cutting required trainings from 8,640 to 150 (~98% reduction) while still finding a configuration that improved accuracy by ~3 points over the untuned baseline.

The most notable result in this run is that **increasing vocabulary size did not help** — 5K features slightly outperformed and 10K. Since the same tree hyperparameters were selected regardless of vocabulary size, the bottleneck appears to be the tree's structural capacity (depth and leaf constraints) rather than the richness of the feature set. This is a useful contrast to linear/probabilistic models, where larger vocabularies more reliably translate into better performance.

---

# 13. Final Conclusion

This experiment evaluated Decision Tree performance on binary text classification using Bag-of-Words features at three vocabulary sizes, with RandomizedSearchCV used for hyperparameter tuning.

The best-evaluated model used **CountVectorizer with 5,000 features and bigrams**, paired with an optimized **DecisionTreeClassifier** (random splitting, Gini impurity, max_depth=20), achieving **74.34% accuracy** and **0.7441 ROC-AUC** — outperforming the untuned baseline by nearly 3 points and edging out the 10K-feature configuration.

Unlike prior experiments, more features did not translate into better performance here, and the 10K configuration showed the more stable cross-validation spread despite a marginally lower mean accuracy. Overall, these results reinforce that single Decision Trees are capacity-bound rather than vocabulary-bound, and remain a weaker choice for high-dimensional sparse text data compared to linear and probabilistic models.